rllib/core/optim/tests/test_rl_optimizer_tf.py

```diff
@@ -0,0 +1,477 @@
+        for key, optim_dict in state.items():
+            self.get_optimizers()[key].set_weights(optim_dict)
+
+    def _configure_optimizers(self) -> Mapping[str, Any]:
+        return {
+            "policy": tf.keras.optimizers.Adam(
+                learning_rate=self._config.get("lr", 1e-3)
+            )
+        }
+
+
+class BCTFTrainer:
+    def __init__(self, config: Mapping[str, any]) -> None:
+        module_config = config["module_config"]
+        optimizer_config = {}
+        self._module = BCTFModule(module_config)
+        self._rl_optimizer = BCTFOptimizer(self._module, optimizer_config)
+
+    @staticmethod
+    def on_after_compute_gradients(
+        gradients_dict: Mapping[str, Any]
+    ) -> Mapping[str, Any]:
+        """Called after gradients have been computed.
+
+        Args:
+            gradients_dict (Mapping[str, Any]): A dictionary of gradients.
+
+        Note the relative order of operations looks like this:
+            fwd_out = forward_train(batch)
+            loss = compute_loss(batch, fwd_out)
+            gradients = compute_gradients(loss)
+            ---> post_processed_gradients = on_after_compute_gradients(gradients)
+            apply_gradients(post_processed_gradients)
+
+        Returns:
+            Mapping[str, Any]: A dictionary of gradients.
+        """
+        return gradients_dict
+
+    def compile_results(
+        self,
+        batch: NestedDict,
+        fwd_out: Mapping[str, Any],
+        postprocessed_loss: Mapping[str, Any],
+        post_processed_gradients: Mapping[str, Any],
+    ) -> Mapping[str, Any]:
+        """Compile results from the update.
+
+        Args:
+            batch: The batch that was used for the update.
+            fwd_out: The output of the forward train pass.
+            postprocessed_loss: The loss after postprocessing.
+            post_processed_gradients: The gradients after postprocessing.
+
+        Returns:
+            A dictionary of results.
+        """
+
+        loss_numpy = convert_to_numpy(postprocessed_loss)
+        results = {
+            "avg_reward": tf.math.reduce_mean(batch["rewards"]).numpy(),
+        }
+        results.update(loss_numpy)
+        return results
+
+    @tf.function
+    def _do_update_fn(self, batch: SampleBatch) -> Mapping[str, Any]:
+        with tf.GradientTape() as tape:
+            # fwd_out = self._module.forward_train(batch)
+            fwd_out = self._module(batch)
+            loss = self._rl_optimizer.compute_loss(batch, fwd_out)
+        gradients = self.compute_gradients(loss, tape)
+        self.apply_gradients(gradients)
+        return {"loss": loss, "fwd_out": fwd_out, "post_processed_gradients": gradients}
+
+    def update(self, batch: SampleBatch) -> Mapping[str, Any]:
+        """Perform an update on this Trainer.
+
+        Args:
+            batch: A batch of data.
+
+        Returns:
+            A dictionary of results.
+        """
+        for key, value in batch.items():
```

Comment:
should we do this here? or before the update call? Sth to think about. The reason is that in training, we have a SampleBatch that has this set_interceptor() thing. I don't like that pattern and prefer to be more explicit like what you did here.

Reply:
hmmm @sven1977 can you chime in here actually? 

do we need to include this interceptor logic during training, in the new RLModule design?
for the purpose of this test we can ignore, but will need an answer when we merge the RLTrainer.